### PR TITLE
FS-1123: Supermicro (SMC) BIOS config Get, Set and Reset support via SUM

### PR DESCRIPTION
**Pull request by splaspood**
## What does this PR implement/change/remove?

BIOS config management support for Supermicro (SMC) platforms implemented as a wrapper over Supermicro Update Manager (SUM).
- GetBiosConfiguration
- SetBiosConfiguration
- SetBiosConfigurationFromFile
- ResetBiosConfiguration

## Description for changelog/release notes
BIOS config management support for Supermicro (SMC) platforms.

```
```

### CODE REVIEW

**Comment by splaspood** [comment]
Still need to complete some code here.

**Comment by mkukiell** [comment]
@splaspood is this ready for review? set reviewers in that case.

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:26]: These should go into the supermicro provider

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:172]: could we have some test cases here that verify the expected command is being executed

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:107]: suggest using Debug or Trace instead of Info

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:99]: add tests

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:133]: add tests

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:236]: The raw prefix is an ironlib thing, can we instead have method in bmc-toolbox that adds and removes this prefix, and that method is invoked here and in ironlib. This will provide more context into this prefix value.

File: providers/sum/sum.go
Comment (joelrebel) [inline providers/sum/sum.go:130]: use implicit returns through this code base

File: providers/sum/sum.go
Comment (splaspood) [reply providers/sum/sum.go]: I believe this logger uses that V(9) to define what level of 'verbosity' something is logged at?    Correct me if I'm wrong...   but there's no .Debug() available off that as-is.

Comment (joelrebel) [reply providers/sum/sum.go]: you're right there, ignore my suggestion

File: .devcontainer/Dockerfile
Comment (DoctorVin) [inline .devcontainer/Dockerfile:1]: why are we using a Microsoft golang container?

File: internal/executor/executor.go
Comment (DoctorVin) [inline internal/executor/executor.go:29]: Do these have to be exported?

File: internal/executor/fake_executor.go
Comment (DoctorVin) [inline internal/executor/fake_executor.go:1]: Why not use `mockery` or `gomock` here?

**Review by DoctorVin — CHANGES_REQUESTED** [review]
I think hand-making your own mocks is not useful here. Using a test-mock generator will allow you to avoid having to expose test infrastructure in the interface and give you better flexibility.

File: .devcontainer/Dockerfile
Comment (splaspood) [reply .devcontainer/Dockerfile:1]: I'd guess simply because that's what's suggested in the devcontainer documentation/examples.     My understanding is there's nothing really special about the devcontainer image(s) so we could replace that with anything.

File: internal/executor/executor.go
Comment (splaspood) [reply internal/executor/executor.go:29]: I'd have to review the code in more detail, but my guess would be these are exported for use in testing.    They are exported here because they were exported in the ironlib code that this was directly copied from.

File: internal/executor/fake_executor.go
Comment (splaspood) [reply internal/executor/fake_executor.go:1]: Only because this already existed in our code base.    I don't have anything against using these libs...

File: internal/executor/executor.go
Comment (jacobweinstock) [inline internal/executor/executor.go:14]: this comment says this is for tests but this interface is used the the `sum` package which is not a test package. which one is it, for tests or not?

Also, why is this a package at all? As far as i can tell this is only used in the `sum` package. This should probably just be move there.

File: internal/sum/sum.go
Comment (jacobweinstock) [inline internal/sum/sum.go:18]: what is `mvcli`?

File: internal/sum/sum.go
Comment (jacobweinstock) [inline internal/sum/sum.go:1]: is `sum` an acronym?

File: internal/executor/executor.go
Comment (splaspood) [reply internal/executor/executor.go:14]: It's external to the sum package with the expectation that there could/would be other similar utilities needing this in the future.

File: internal/sum/sum.go
Comment (splaspood) [reply internal/sum/sum.go]: good catch, thank you.

File: internal/sum/sum.go
Comment (splaspood) [reply internal/sum/sum.go:1]: Yes, Supermicro Update Manager https://www.supermicro.com/en/solutions/management-software/supermicro-update-manager

File: internal/executor/executor.go
Comment (jacobweinstock) [reply internal/executor/executor.go:14]: what the the other utilities that will need and or implement this? if we dont have other uses we should hold off creating this.

This also is a big interface and its purpose and value add are not very clear. are we planning on replacing all calls to the `os/exec` with this? 

"Do not define interfaces before they are used: without a realistic example of usage, it is too difficult to see whether an interface is even necessary, let alone what methods it ought to contain." - https://go.dev/wiki/CodeReviewComments#interfaces

File: internal/sum/sum.go
Comment (jacobweinstock) [reply internal/sum/sum.go:1]: thanks, please make a package level code comment so we dont lose this info once the PR is merged.

File: internal/sum/sum.go
Comment (jacobweinstock) [inline internal/sum/sum.go:77]: Any reason this can't live in a test package only?

File: internal/executor/executor.go
Comment (splaspood) [reply internal/executor/executor.go:14]: I do not have any currently pending, so your argument about not implementing it before it's necessary is valid.   

That being said, this code was already implemented elsewhere for a very similar use case so it was adopted nearly as-is.    The presumption is that we'd need the currently unused functionality as we add more vendor utilities as we've seen in ironlib.

Comment (jacobweinstock) [reply internal/executor/executor.go:14]: im not familiar with the ironlib code base. does ironlib already have the supermicro update manager code? should we just be importing from ironlib?

I'm also confused. you say "do not have any currently pending" and "as we add more vendor utilities". How can this be both?

File: internal/sum/sum.go
Comment (splaspood) [reply internal/sum/sum.go]: Nope, and you're correct it's not used anywhere.   Bye bye.

File: internal/executor/executor.go
Comment (splaspood) [reply internal/executor/executor.go:14]: Ironlib has many things, but not sum.   I was referring to the Executor/FakeExecutor stuff that's used for testing/execing in ironlib.

While I do not currently have any vendor utilities in the queue, I can envision a future with supporting other vendor's utilities where this code would be reused.

File: providers/sum/sum.go
Comment (splaspood) [reply providers/sum/sum.go]: I'd like to handle this in a future PR.

**Review by joelrebel — APPROVED** [review]

File: internal/executor/executor.go
Comment (joelrebel) [reply internal/executor/executor.go:14]: > should we just be importing from ironlib?
ironlib is mean't to execute vendor utilities in an OS running on the host.
bmclib is mean't to execute actions/calls remotely and so I think this fits in here

File: internal/executor/executor.go
Comment (joelrebel) [reply internal/executor/executor.go:29]: when we're using a mock library, these would not be required

File: internal/executor/executor.go
Comment (splaspood) [reply internal/executor/executor.go:14]: If we plan to continue using this code over the longer term vs something off the shelf as suggested by @DoctorVin , I'd argue that this should be extracted into an external package that both projects could import, but we haven't finalized those plans one way or the other at this point.

**Pull request merged**
